File: day 2/main.py
def puzzleOne():
    print("\n Puzzle 1: \n")

    with open("day 2/input.txt") as input:
        answer = 0

        for line in input:
            # we are going to break the string appart in multiple "steps". Python's built-in split function will help out greatly here

            # the first one will be separating the policy—number of minimum and maximum appearences—from the password itself
            step1 = line.split(":")
            password = step1[1]

            # next step is to seperate the number of occurences from the letter in the policy section
            step2 = step1[0].split(" ")
            letter = step2[1]

            # last step is to get the min and max from the step2 result
            step3 = step2[0].split("-")
            min, max = step3[0], step3[1]

            # then we check if the number of occurences of the letter is bewteen the min and max inclusive, if so, add one to answer
            if password.count(letter) >= int(min) and password.count(letter) <= int(
                max
            ):
                answer += 1

        print("puzzle one: ", answer)


def puzzleTwo():
    print("\n Puzzle 2: \n")

    with open("day 2/input.txt") as input:
        answer = 0

        for line in input:
            # again, we will be using multiple "steps" to seperate the different parts of the input strings

            # separate the policy from the password
            step1 = line.split(":")
            password = step1[1].strip()

            # separate the plocy from the character
            step2 = step1[0].split(" ")
            letter = step2[1]

            # seperate the first location from the second location
            step3 = step2[0].split("-")
            # we have to remove one from the first and second locations because array indexing starts at 0, but the counting used for the policy creation starts at 1
            first = int(step3[0]) - 1
            second = int(step3[1]) - 1

            # XOR check for the positional chracters in the password
            if password[first] == letter and password[second] != letter:
                answer += 1

            if password[first] != letter and password[second] == letter:
                answer += 1

        print("puzzle two: ", answer)


puzzleOne()

puzzleTwo()

File: day 2/main.rs
use std::fs::File;
use std::io::*;

fn main() {
    puzzle_one();
    puzzle_two();
}

fn count_characters(input: &str, character: &char) -> usize {
    return input.matches(*character).count();
}

fn puzzle_one() {
    println!("\n Puzzle 1: \n");

    let mut valid_passwords_count: usize = 0;

    let input = File::open("./input.txt").unwrap();
    let reader = BufReader::new(input);

    let mut range: &str;
    let mut character_string: &str;
    let mut password: &str;

    for line in reader.lines() {
        let line_value = line.unwrap();
        let mut parts_iter = line_value.split_whitespace();

        range = parts_iter.next().unwrap();
        character_string = &parts_iter.next().unwrap()[0..1];
        password = parts_iter.next().unwrap();

        let character = character_string.chars().nth(0).unwrap();

        let num_chars: usize = count_characters(password, character);

        let mut range_iter = range.split("-");

        let range_min: usize =
            range_iter.next().unwrap().parse::<usize>().unwrap();
        let range_max: usize =
            range_iter.next().unwrap().parse::<usize>().unwrap();

        if range_min <= num_chars && num_chars <= range_max {
            valid_passwords_count += 1;
        }
    }

    println!("{}", valid_passwords_count);
}

fn puzzle_two() {
    println!("\n Puzzle 2: \n");

    let mut valid_passwords_count: usize = 0;

    let input = File::open("./input.txt").unwrap();
    let reader = BufReader::new(input);

    for line in reader.lines() {
        let line_value = line.unwrap();
        let mut parts_iter = line_value.split_whitespace();

        let range = parts_iter.next().unwrap();
        let character_string = &parts_iter.next().unwrap()[0..1];
        let password = parts_iter.next().unwrap();

        let character_array = character_string.chars().collect::<Vec<_>>();
        let character = character_array.get(0).unwrap();

        let mut range_iter = range.split("-");

        let pos_one: usize =
            range_iter.next().unwrap().parse::<usize>().unwrap();
        let pos_two: usize =
            range_iter.next().unwrap().parse::<usize>().unwrap();

        let pos_one_char = password.chars().nth(pos_one - 1).unwrap();
        let pos_two_char = password.chars().nth(pos_two - 1).unwrap();

        if (pos_one_char == *character) ^ (pos_two_char == *character) {
            valid_passwords_count += 1;
        }
    }
    println!("{}", valid_passwords_count);
}
